v2/lints/mozilla/lint_mp_pss_parameters_encoding_correct.go

```diff
@@ -0,0 +1,98 @@
+package mozilla
+
+/*
+ * ZLint Copyright 2018 Regents of the University of Michigan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+/************************************************
+
+https://www.mozilla.org/en-US/about/governance/policies/security-group/certs/policy/
+
+Section 5.1.1 RSA
+
+RSASSA-PSS with SHA-256, MGF-1 with SHA-256, and a salt length of 32 bytes.
+
+The encoded AlgorithmIdentifier MUST match the following hex-encoded bytes:
+
+304106092a864886f70d01010a3034a00f300d0609608648016503040201
+0500a11c301a06092a864886f70d010108300d0609608648016503040201
+0500a203020120
+
+RSASSA-PSS with SHA-384, MGF-1 with SHA-384, and a salt length of 48 bytes.
+
+The encoded AlgorithmIdentifier MUST match the following hex-encoded bytes:
+
+304106092a864886f70d01010a3034a00f300d0609608648016503040202
+0500a11c301a06092a864886f70d010108300d0609608648016503040202
+0500a203020130
+
+RSASSA-PSS with SHA-512, MGF-1 with SHA-512, and a salt length of 64 bytes.
+
+The encoded AlgorithmIdentifier MUST match the following hex-encoded bytes:
+
+304106092a864886f70d01010a3034a00f300d0609608648016503040203
+0500a11c301a06092a864886f70d010108300d0609608648016503040203
+0500a203020140
+************************************************/
+
+import (
+	"bytes"
+	"encoding/hex"
+
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/v2/lint"
+	"github.com/zmap/zlint/v2/util"
+)
+
+type rsaPssAidEncoding struct{}
+
+var allowedAidEncodingsForPSS = [3]string{
+	"304106092a864886f70d01010a3034a00f300d06096086480165030402010500a11c301a06092a864886f70d010108300d06096086480165030402010500a203020120",
+	"304106092a864886f70d01010a3034a00f300d06096086480165030402020500a11c301a06092a864886f70d010108300d06096086480165030402020500a203020130",
+	"304106092a864886f70d01010a3034a00f300d06096086480165030402030500a11c301a06092a864886f70d010108300d06096086480165030402030500a203020140"}
+
```

Comment:
I don't think it makes sense to store these values as strings of hex encoded bytes. It means you have to decode them again in `Execute` for each certificate that's linted. It would be better to express these values as a `[]byte`. I also think you'll want to express this as a map from OID to expected `[]byte` (see my later comments in `Execute` about the `for` loop based comparison).

This would also match how the `CheckAlgorithmIDParamNotNULL` function used by the `e_tbs_signature_rsa_encryption_parameter_not_null` lint is implemented: https://github.com/zmap/zlint/blob/5b2df5c915f1d73dbfcd4f2ac2a6241103fbc4a3/v2/util/algorithm_identifier.go#L14-L32

Reply:
I am not sure about the map approach. In the case of RSAPSS this would look something like that:

var RSASSAPSSAlgorithmIDAndParametersToDER = map[string][]byte{
	"RSASSA-PSS with SHA-256, MGF-1 with SHA-256, salt length 32 bytes": {0x30, 0x41, .... 0xa2, 0x03, 0x02, 0x01, 0x20}
}

because there is no direct mapping of an OID to DER encoding. The OID for PSS is always the same. I will change this to working with bytes instead of hex encoded strings. Ok?

Reply:
Changed to map for better readability, but the key of the map is not used.

Reply:
> because there is no direct mapping of an OID to DER encoding. The OID for PSS is always the same.

Ahhhhhh! I understand now. Apologies, I mistakenly thought the OID would change between these cases and could be used to index the expected bytes for the DER encoding of the parameters. Thanks for your patience/explanation.

I think the best solution in this case is a hybrid between what you have now (a `map[string][]byte`) and what you had originally (a `[3]string`). Since the `map` keys aren't used anywhere in the code I think they're better expressed as comments. I would like to see the `[]byte` constants remain the same. So I think the best approach will be using `[3][]byte` and adding comments on each constant. E.g. something like:

```go
var RSASSAPSSAlgorithmIDToDER = [3][]byte{
        // RSASSA-PSS with SHA-256, MGF-1 with SHA-256, salt length 32 bytes
	{0x30, 0x41, 0x06, 0x09, 0x2a, 0x86, 0x48, 0x86, 0xf7, 0x0d, 0x01, 0x01, 0x0a, 0x30, 0x34, 0xa0, 0x0f, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x01, 0x05, 0x00, 0xa1, 0x1c, 0x30, 0x1a, 0x06, 0x09, 0x2a, 0x86, 0x48, 0x86, 0xf7, 0x0d, 0x01, 0x01, 0x08, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x01, 0x05, 0x00, 0xa2, 0x03, 0x02, 0x01, 0x20},
        // RSASSA-PSS with SHA-384, MGF-1 with SHA-384, salt length 48 bytes
	{0x30, 0x41, 0x06, 0x09, 0x2a, 0x86, 0x48, 0x86, 0xf7, 0x0d, 0x01, 0x01, 0x0a, 0x30, 0x34, 0xa0, 0x0f, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x02, 0x05, 0x00, 0xa1, 0x1c, 0x30, 0x1a, 0x06, 0x09, 0x2a, 0x86, 0x48, 0x86, 0xf7, 0x0d, 0x01, 0x01, 0x08, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x02, 0x05, 0x00, 0xa2, 0x03, 0x02, 0x01, 0x30},
        // RSASSA-PSS with SHA-512, MGF-1 with SHA-512, salt length 64 bytes
	{0x30, 0x41, 0x06, 0x09, 0x2a, 0x86, 0x48, 0x86, 0xf7, 0x0d, 0x01, 0x01, 0x0a, 0x30, 0x34, 0xa0, 0x0f, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x03, 0x05, 0x00, 0xa1, 0x1c, 0x30, 0x1a, 0x06, 0x09, 0x2a, 0x86, 0x48, 0x86, 0xf7, 0x0d, 0x01, 0x01, 0x08, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x03, 0x05, 0x00, 0xa2, 0x03, 0x02, 0x01, 0x40},
}
```

Reply:
Ok.